com.sap.cloud.lm.sl.cf.process/src/test/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineDesiredStateAchievingActionsStepTest.java

```diff
@@ -1,177 +1,420 @@
+        boolean hasAppChanged, Set<ApplicationStateAction> expectedAppStateActions, List<CloudBuild> cloudBuilds,
+        Class<? extends Exception> exception) {
+        super(currentAppState, desiredAppState, hasAppChanged, expectedAppStateActions, cloudBuilds);
+        this.exception = exception;
     }
 
     @Test
     public void testExecute() throws Exception {
+        if (exception != null) {
+            expectedException.expect(exception);
+        }
         step.execute(context);
-
         assertStepFinishedSuccessfully();
 
         assertEquals(expectedAppStateActions, StepsUtil.getAppStateActionsToExecute(context));
     }
 
     @Override
-    protected DetermineDesiredStateAchievingActionsStep createStep() {
-        return new DetermineDesiredStateAchievingActionsStep();
+    protected void initProperties() {
+        context.setVariable(Constants.VAR_VCAP_APP_PROPERTIES_CHANGED, false);
+        context.setVariable(Constants.VAR_VCAP_SERVICES_PROPERTIES_CHANGED, false);
+        context.setVariable(Constants.VAR_USER_PROPERTIES_CHANGED, false);        
     }
 
+    @Override
+    protected RestartParameters getRestartParameters() {
+        return new RestartParameters(false, false, false);
+    }
+
+    @RunWith(Parameterized.class)
+    public static class DetermineAppRestartTest extends DetermineDesiredStateAchievingActionsStepBaseTest {
+
+
+        @Parameters
+        public static List<Object[]> getParameters() throws ParseException {
+            return Arrays.asList(new Object[][] {
+                // @formatter:off
+                // (0)
+                {
+                    ApplicationStartupState.STARTED, ApplicationStartupState.STARTED, false, Stream.of(ApplicationStateAction.STAGE, ApplicationStateAction.START, ApplicationStateAction.STOP).collect(Collectors.toSet()), Collections.emptyList(), true, true, true, true, true, true
+                },
+                // (1)
+                {
+                    ApplicationStartupState.STARTED, ApplicationStartupState.STARTED, false, Stream.of(ApplicationStateAction.STAGE, ApplicationStateAction.START, ApplicationStateAction.STOP).collect(Collectors.toSet()), Collections.emptyList(), true, false, false, true, false, false
+                },
+                // (2)
+                {
+                    ApplicationStartupState.STARTED, ApplicationStartupState.STARTED, false, Stream.of(ApplicationStateAction.STAGE, ApplicationStateAction.START, ApplicationStateAction.STOP).collect(Collectors.toSet()), Collections.emptyList(), false, true, false, false, true, false
+                },
+                // (3)
+                {
+                    ApplicationStartupState.STARTED, ApplicationStartupState.STARTED, false, Stream.of(ApplicationStateAction.STAGE, ApplicationStateAction.START, ApplicationStateAction.STOP).collect(Collectors.toSet()), Collections.emptyList(), false, false, true, false, false, true
+                },
+                // (4)
```

Comment:
Add a comment explaining why the expected actions always contain `ApplicationStateAction.STAGE`.